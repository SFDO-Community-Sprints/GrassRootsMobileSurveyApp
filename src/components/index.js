export * from './TextHeading';
export * from './TextSubheading';
export * from './ListItem';
export * from './Loader';
export * from './TextIcon';
export * from './LoginFormInput';
