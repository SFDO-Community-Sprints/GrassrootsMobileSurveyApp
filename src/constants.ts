import { ImageStyle, ViewStyle, Dimensions } from 'react-native';
import { SQLiteFieldTypeMapping } from './types/sqlite';
import { Language } from './types/metadata';

export const APP_FONTS = {
  FONT_REGULAR: 'SalesforceSans-Regular',
  FONT_BOLD: 'SalesforceSans-Bold',
  FONT_LIGHT: 'SalesforceSans-Light',
};

export const APP_THEME = {
  APP_BASE_COLOR: '#01579b',
  NAVIGATION_BACKGROUND: '#f8f8f8',
  APP_BASE_FONT_COLOR: '#01579b',
  APP_DARK_FONT_COLOR: '#2b2826',
  APP_LIGHT_FONT_COLOR: '#706e6b',
  APP_BORDER_COLOR: '#dedcdb',
  APP_ERROR_COLOR: '#c23934',
  APP_LIST_HEADER_COLOR: '#e1f5fe',
  APP_WHITE: 'white',
  APP_SUCCESS_COLOR: '#04844b',
  APP_GRAY_BACKGROUND_COLOR: '#f3f2f2',
  APP_DISABLED_BACKGROUND_COLOR: '#c9c7c5',
};

export const ASYNC_STORAGE_KEYS = {
  SALESFORCE_INSTANCE_URL: '@SalesforceInstanceURL',
  FIELD_WORKER_CONTACT_ID: '@FieldWorkerContactId',
  FIELD_TYPE: '@FieldType',
  LANGUAGES: '@Languages',
};

export const SECURE_STORE_KEYS = {
  EMAIL: 'email',
  PASSWORD: 'password',
  ACCESS_TOKEN: 'token',
  HEROKU_APP_NAME: 'heroku',
};

export const BACKGROUND_IMAGE_SOURCE = require('../assets/images/background.png');
const ratio = Dimensions.get('window').width / 1050; // 1050 is image width
const height = 534 * ratio;

// Style for background container
export const BACKGROUND_STYLE: ViewStyle = {
  flex: 1,
  backgroundColor: APP_THEME.APP_WHITE,
};

// Style for background image itself
export const BACKGROUND_IMAGE_STYLE: ImageStyle = {
  resizeMode: 'contain',
  top: Dimensions.get('window').height / 2 + height / 2,
};

export const SURVEY_OBJECT = 'GRMS__Survey__c';
export const FIELD_WORKER_CONTACT_FIELD_ON_SURVEY = 'GRMS__FieldWorker__c';
export const SURVEY_DATE_FIELD = 'GRMS__SurveyDate__c';
export const AVAILABLE_LANGUAGE_CMDT = 'GRMS__AvailableLanguage__mdt';
export const RECORD_TYPE_ID_FIELD = 'RecordTypeId';
export const BACKGROUND_SURVEY_FIELDS = [
  { fieldName: 'Name', fieldType: 'string' },
  { fieldName: RECORD_TYPE_ID_FIELD, fieldType: 'reference' },
  { fieldName: FIELD_WORKER_CONTACT_FIELD_ON_SURVEY, fieldType: 'reference' },
  { fieldName: SURVEY_DATE_FIELD, fieldType: 'datetime' },
];
export const LOCAL_SURVEY_FIELDS: Array<SQLiteFieldTypeMapping> = [{ name: '_syncStatus', type: 'text' }];

export const SYNC_STATUS_SYNCED = 'Synced';
export const SYNC_STATUS_UNSYNCED = 'Unsynced';

export const SUBSCRIBER_PACKAGE_VERSION = '0336g000000kHHoAAM';
export const MIN_PACKAGE_VERSION = '0.4.0.5';

export const DB_TABLE = {
  RECORD_TYPE: 'RecordType',
  PAGE_LAYOUT_SECTION: 'PageLayoutSection',
  PAGE_LAYOUT_ITEM: 'PageLayoutItem',
  PICKLIST_VALUE: 'PicklistValue',
  SURVEY: 'Survey',
  LOCALIZATION: 'Localization',
  CONTACT: 'Contact',
};

export const L10N_PREFIX = {
  RecordType: 'RECORD_TYPE_',
  PageLayoutSectionId: 'PAGE_LAYOUT_SECTION_',
  PageLayoutItem: 'PAGE_LAYOUT_ITEM_',
};

export const DEFAULT_SF_LANGUAGE: Language = {
  name: 'English',
  code: 'en_US',
};
export const SUPPORTED_SF_LANGUAGES: Array<Language> = [
  { name: 'Chinese (Simplified)', code: 'zh_CN' },
  { name: 'Chinese (Traditional)', code: 'zh_TW' },
  { name: 'Danish', code: 'da' },
  { name: 'Dutch', code: 'nl_NL' },
  DEFAULT_SF_LANGUAGE,
  { name: 'Finnish', code: 'fi' },
  { name: 'French', code: 'fr' },
  { name: 'German', code: 'de' },
  { name: 'Italian', code: 'it' },
  { name: 'Japanese', code: 'ja' },
  { name: 'Korean', code: 'ko' },
  { name: 'Norwegian', code: 'no' },
  { name: 'Portuguese (Brazil)', code: 'pt_BR' },
  { name: 'Russian', code: 'ru' },
  { name: 'Spanish', code: 'es' },
  { name: 'Spanish (Mexico)', code: 'es_MX' },
  { name: 'Swedish', code: 'sv' },
  { name: 'Thai', code: 'th' },
  { name: 'Nepali', code: 'ne' }, // For HaydenHall ‚ù§Ô∏èüá≥üáµ
];
