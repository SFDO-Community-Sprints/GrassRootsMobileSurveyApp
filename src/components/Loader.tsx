import React from 'react';
import { StyleSheet, View, Modal, ActivityIndicator, Text } from 'react-native';
import { APP_THEME, APP_FONTS } from '../constants';

type LoaderProps = {
  loading: boolean;
  message?: string;
};

function Loader(props: LoaderProps) {
  const { loading, message } = props;
  return (
    <Modal transparent visible={loading}>
      <View style={styles.modalBackground}>
        <ActivityIndicator animating={loading} size="large" hidesWhenStopped color={APP_THEME.APP_BASE_COLOR} />
        <Text style={styles.loadingTitleStyle}> {message} </Text>
      </View>
    </Modal>
  );
}

const styles = StyleSheet.create({
  modalBackground: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: APP_THEME.APP_WHITE,
    alignItems: 'center',
    opacity: 0.8,
  },
  loadingTitleStyle: {
    fontFamily: APP_FONTS.FONT_REGULAR,
    fontSize: 36,
    color: APP_THEME.APP_BASE_COLOR,
    paddingTop: 10,
  },
});

export { Loader };
