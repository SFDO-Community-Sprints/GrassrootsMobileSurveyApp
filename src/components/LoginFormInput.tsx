import React from 'react';
import { StyleSheet } from 'react-native';
import { Input } from '@rneui/themed';
import { APP_THEME, APP_FONTS } from '../constants';

type LoginFormInputProps = {
  onChangeText(value: string): void;
  value: string;
  label: string;
  placeholder: string;
  icon: string;
  isSecure?: boolean;
  errorMessage: string;
  keyboardType?: 'email-address' | 'default';
};

function LoginFormInput(props: LoginFormInputProps) {
  return (
    <Input
      onChangeText={value => {
        props.onChangeText(value);
      }}
      value={props.value}
      label={props.label}
      placeholder={props.placeholder}
      leftIcon={{ type: 'material-community', name: props.icon, color: APP_THEME.APP_LIGHT_FONT_COLOR }}
      errorStyle={styles.errorStyle}
      labelStyle={styles.font}
      inputStyle={styles.font}
      keyboardType={props.keyboardType}
      errorMessage={props.errorMessage}
      secureTextEntry={props.isSecure}
    />
  );
}

const styles = StyleSheet.create({
  font: {
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
  errorStyle: {
    color: APP_THEME.APP_ERROR_COLOR,
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
});

export { LoginFormInput };
