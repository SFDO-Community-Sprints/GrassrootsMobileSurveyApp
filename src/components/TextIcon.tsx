import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { Icon } from '@rneui/themed';
import { APP_FONTS } from '../constants';

type TextIconPropType = {
  icon: string;
  color: string;
  children: React.ReactNode;
};

function TextIcon(props: TextIconPropType) {
  return (
    <View style={styles.container}>
      <Icon iconStyle={{ padding: 10 }} name={props.icon} size={22} color={props.color} type="material" />
      <Text style={styles.textStyle}>{props.children}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    padding: 10,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
  },
  textStyle: {
    fontSize: 16,
    lineHeight: 22,
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
});

export { TextIcon };
