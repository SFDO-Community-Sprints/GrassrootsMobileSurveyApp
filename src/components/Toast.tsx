import React from 'react';
import { BaseToast, ErrorToast } from 'react-native-toast-message';
import { APP_THEME } from '../constants';

/**
 * @description Toast config using SLDS color
 */
export const toastConfig = {
  /*
    Overwrite 'success' type,
    by modifying the existing `BaseToast` component
  */
  success: props => (
    <BaseToast
      {...props}
      style={{ borderLeftColor: APP_THEME.APP_SUCCESS_COLOR }}
      text1Style={{
        fontSize: 15,
      }}
      text2Style={{
        fontSize: 12,
      }}
    />
  ),
  /*
    Overwrite 'error' type,
    by modifying the existing `ErrorToast` component
  */
  error: props => (
    <ErrorToast
      {...props}
      style={{ borderLeftColor: APP_THEME.APP_ERROR_COLOR }}
      text1Style={{
        fontSize: 15,
      }}
      text2Style={{
        fontSize: 12,
      }}
    />
  ),
};
