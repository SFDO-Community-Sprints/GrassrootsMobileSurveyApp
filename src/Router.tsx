import React, { useEffect, useState } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';

import { useLocalizationContext } from './context/localizationContext';
import { AuthContextValue, useAuthContext } from './context/authContext';

// screens
import Login from './screens/Login';
import LoginSettings from './screens/LoginSettings';
import SurveyList from './screens/SurveyList';
import SurveyTypePicker from './screens/SurveyTypePicker';
import SurveyEditor from './screens/SurveyEditor';
import Settings from './screens/Settings';
import LookupSearch from './screens/LookupSearch';

// components
import { LogoutButton } from './components/surveyList/SurveyListHeaderButtons';
import { LoginSettingsButton } from './components/LoginSettingsButton';

// styles
import { ASYNC_STORAGE_KEYS, APP_FONTS, APP_THEME } from './constants';

export type StackParamList = {
  Login: undefined;
  LoginSettings: undefined;
  SurveyList: undefined;
  SurveyTypePicker: undefined;
  SurveyEditor: {
    selectedRecordTypeId?: string;
    selectedLayoutId?: string;
    _localId?: string;
  };
  Settings: undefined;
  Lookup: {
    fieldName: string;
    title: string;
  };
};

const MainStack = createStackNavigator();
const RootStack = createStackNavigator();
const LoginStack = createStackNavigator();

function LoginStackScreen() {
  const { t } = useLocalizationContext();
  return (
    <LoginStack.Navigator>
      <MainStack.Screen
        name="Login"
        component={Login}
        options={({ navigation }) => ({
          title: t('LOGIN'),
          headerRight: () => LoginSettingsButton({ navigation }),
          ...headerStyle,
          animationEnabled: false,
        })}
      />
      <MainStack.Screen
        name="LoginSettings"
        component={LoginSettings}
        options={{
          title: t('LOGIN_SETTINGS'),
          ...headerStyle,
        }}
      />
    </LoginStack.Navigator>
  );
}

function MainStackScreen() {
  const { t } = useLocalizationContext();
  const authContext = useAuthContext();
  return (
    <MainStack.Navigator>
      <MainStack.Screen
        name="SurveyList"
        component={SurveyList}
        options={() => ({
          title: t('SURVEYS'),
          headerLeft: () => LogoutButton(t, authContext),
          ...headerStyle,
          gestureEnabled: false,
          animationEnabled: false,
        })}
      />
      <MainStack.Screen
        name="SurveyTypePicker"
        component={SurveyTypePicker}
        options={{
          title: t('CHOOSE_SURVEY'),
          headerBackTitle: 'Back',
          ...headerStyle,
        }}
      />
      <MainStack.Screen
        name="SurveyEditor"
        component={SurveyEditor}
        options={{
          title: t('NEW_SURVEY'),
          headerBackTitle: 'Back',
          ...headerStyle,
        }}
      />
      <MainStack.Screen
        name="Settings"
        component={Settings}
        options={{
          title: t('SETTINGS'),
          headerBackTitle: 'Back',
          ...headerStyle,
        }}
      />
    </MainStack.Navigator>
  );
}

export default function Router() {
  const authContext: AuthContextValue = useAuthContext();
  const [loaded, setLoaded] = useState<boolean>(false);

  useEffect(() => {
    const prepare = async () => {
      try {
        const fieldWorkerContactId = await storage.load({ key: ASYNC_STORAGE_KEYS.FIELD_WORKER_CONTACT_ID });
        if (fieldWorkerContactId) {
          authContext.login();
        }
      } catch {
        // storage.load fails when no data
      } finally {
        setLoaded(true);
      }
    };
    prepare();
  }, []);

  return loaded ? (
    <NavigationContainer>
      <RootStack.Navigator mode="modal" headerMode="none">
        {authContext.isLoggedIn ? (
          <RootStack.Screen name="Main" component={MainStackScreen} />
        ) : (
          <RootStack.Screen name="Unauthed" component={LoginStackScreen} />
        )}
        <RootStack.Screen name="Lookup" component={LookupSearch} />
      </RootStack.Navigator>
    </NavigationContainer>
  ) : null;
}

const headerStyle = {
  headerStyle: {
    backgroundColor: APP_THEME.NAVIGATION_BACKGROUND,
  },
  headerTintColor: APP_THEME.APP_BASE_FONT_COLOR,
  headerTitleStyle: {
    fontSize: 16,
    fontWeight: '600' as '600',
    fontFamily: APP_FONTS.FONT_BOLD,
  },
};
