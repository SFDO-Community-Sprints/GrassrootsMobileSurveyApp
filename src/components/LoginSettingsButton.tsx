import React from 'react';
import { Icon } from '@rneui/themed';
import { StackNavigationProp } from '@react-navigation/stack';

import { APP_THEME } from '../constants';

import { StackParamList } from '../Router';
type LoginScreenNavigationProp = StackNavigationProp<StackParamList, 'SurveyEditor'>;

type LoginSettingsButtonProps = {
  navigation: LoginScreenNavigationProp;
};

export function LoginSettingsButton(props: LoginSettingsButtonProps) {
  return (
    <Icon
      accessibilityRole="button"
      accessible
      iconStyle={{ padding: 10 }}
      name="settings"
      size={22}
      color={APP_THEME.APP_BASE_COLOR}
      type="material"
      onPress={() => {
        props.navigation.navigate('LoginSettings');
      }}
    />
  );
}
