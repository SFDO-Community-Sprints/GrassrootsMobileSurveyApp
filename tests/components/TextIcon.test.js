import React from 'react';
import { render } from '@testing-library/react-native';

import { TextIcon } from '../../src/components';

describe('<TextIcon />', () => {
  it('Rendering', () => {
    const screen = render(
      <TextIcon icon="check-circle" color="#000000">
        Hello
      </TextIcon>
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
