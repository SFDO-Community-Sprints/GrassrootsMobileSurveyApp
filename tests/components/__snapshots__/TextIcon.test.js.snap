// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextIcon /> Rendering 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "flex-start",
      "padding": 10,
    }
  }
>
  <View
    color="#000000"
    iconStyle={
      Object {
        "padding": 10,
      }
    }
    name="check-circle"
    size={22}
    type="material"
  />
  <Text
    style={
      Object {
        "fontFamily": "SalesforceSans-Regular",
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
  >
    Hello
  </Text>
</View>
`;
